scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py

```diff
@@ -67,20 +66,40 @@ def _makeMatMat(m):
         return lambda v: m @ v
 
 
+def _dot_inplace(x, y, verbosityLevel=0):
+    """Perform 'np.dot' in-place if dtypes match."""
+    if y.shape[0] == y.shape[1]:
+        try:
+            np.dot(x, y, out=x)
+        except Exception:
+            if verbosityLevel:
+                warnings.warn(
+                    f"Trying in-place x = x @ y, possibly x.dtype {x.dtype} "
+                    f"does not match dtype of x @ y since y.dtype {y.dtype}, "
+                    f"or x is a slice rather than a C-Array, "
+                    f"so x needs to be overwritten preventing in-place.",
+                    UserWarning, stacklevel=3
+                )
+            x = x @ y
+    else:
+        x = x @ y
+    return x
```

Comment:
Logically, I would prefer to express `y.shape[0] == y.shape[1]` differently. Even though it is correct, it is not obvious IMO why y would need to be square matrix so that `x @ y` has the same shape as `x`.

I find `x.shape[1] == y.shape[1]` to be a clearer expression of this intent (but of course `x.shape[1] == y.shape[0]` is necessary to permit matrix-multiplication in the first place, so it ends up being equivalent to your formulation).

However, the bigger point is that  the `try/except` fallback needs to be in the `else:` branch. The point is that we can formulate sufficient conditions to _ensure_ inplace updates work. When our sufficient conditions fail (rather than exhaustively working through which conditions are _required_), we just try whether an inplace update works and if not fall back to full multiplication.

(This time it is meant as a committable suggestion)

```suggestion
    if x.flags["OWNDATA"] and x.shape[1] == y.shape[1] and x.dtype == y.dtype:
        # conditions where we can guarantee that inplace updates will work;
        # i.e. x is not a view/slice, x & y have compatible dtypes, and the
        # shape of the result of x @ y matches the shape of x.
        np.dot(x, y, out=x)
    else:
        # ideally, we'd have an exhaustive list of conditions above when
        # inplace updates are possible; since we don't, we opportunistically
        # try if it works, and fall back to overwriting if necessary
        try:
            np.dot(x, y, out=x)
        except Exception:
            if verbosityLevel:
                warnings.warn(
                    "Inplace update of x = x @ y failed, "
                    "x needs to be overwritten.",
                    UserWarning, stacklevel=3
                )
            x = x @ y
    return x
```